drivers/serial/uart_stm32.c

```diff
@@ -724,14 +1208,58 @@ static int uart_stm32_init(struct device *dev)
 	}
 #endif /* !USART_ISR_REACK */
 
-#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+#if defined(CONFIG_UART_INTERRUPT_DRIVEN) || defined(CONFIG_UART_ASYNC_API)
 	config->uconf.irq_config_func(dev);
 #endif
-	return 0;
+
+	int ret = 0;
+#ifdef CONFIG_UART_ASYNC_API
+	ret = uart_stm32_async_init(dev);
```

Comment:
Please make sure that it will be possible to have mixed use of async and non async uart devices.

Reply:
I think once both CONFIG_UART_ASYNC_API and CONFIG_UART_INTERRUPT_DRIVEN are enabled, every UART port will have both capabilities. 

I tried the build without CONFIG_UART_ASYNC_API enabled, no complication errors. anyway, I'll check the conditions time to time to make sure it doesn't break.